bam_qc_metrics/bam_qc.py

```diff
@@ -748,20 +749,59 @@ def evaluate_bedtools_metrics(self):
             metrics['number of targets'] = targetBedTool.count()
             metrics['total target size'] = sum(len(f) for f in targetBedTool.features())
             metrics['reads on target'] = len(bamBedTool.intersect(self.target_path))
-            self.logger.debug("Found bedtools metrics")
+            result = self.evaluate_bedtools_coverage(bamBedTool, targetBedTool)
+            [metrics['total coverage'], metrics['coverage per target']] = result
+            self.logger.info("Found bedtools metrics")
         else:
             metrics['number of targets'] = None
             metrics['total target size'] = None
             metrics['reads on target'] = None
+            metrics['total coverage'] = None
+            metrics['coverage per target'] = {}
             self.logger.info("No target path given, omitting bedtools metrics")
-        # placeholders; TODO implement these metrics
-        metrics['total coverage'] = None
-        metrics['coverage per target'] = None
-        # TODO add bedtools coverage metrics?
-        #coverage = targetBedTool.coverage(self.bam_path)
-        #print(coverage)
         return metrics
 
+    def evaluate_bedtools_coverage(self, bamBedTool, targetBedTool):
+        """
+        Find the bedtools coverage histogram and process to get coverage by target.
+        For each target, coverage = total bases on target / target size
+        """
+        hist_str = self.run_bedtools()
+        self.logger.debug("Found bedtools coverage histogram")
+        reader = csv.reader(re.split("\n", hist_str.strip()), delimiter="\t")
+        all_name = 'all'
+        by_target = {}
+        target_sizes = {}
+        # row either starts with 'all', or with (chromosome, start, end)
+        # row ends with (depth, bases, target_size, fraction_covered)
+        int_expr = re.compile("^[0-9]+$")
+        for row in reader:
+            if len(row) == 5 and row[0] == all_name:
+                target = all_name
+            elif len(row) >= 7 and int_expr.match(row[1]) and int_expr.match(row[2]):
+                target = ','.join(row[0:3])
+            else:
+                msg = "Cannot parse target in bedtools output: '"+str(row)+"'"
+                self.logger.error(msg)
+                raise ValueError(msg)
+            coverage_fields = row[-4:]
+            try:
+                [depth, bases, target_size] = [int(x) for x in coverage_fields[0:3]]
+                # coverage_fields[3] = fraction covered; not used here
+            except ValueError:
+                msg = "Cannot parse coverage in bedtools output: '"+str(row)+"'"
+                self.logger.error(msg)
+                raise
+            target_sizes[target] = target_size
+            by_target[target] = by_target.get(target, 0) + depth*bases
+        for target in target_sizes.keys():
+            coverage = float(by_target[target]) / target_sizes[target]
+            by_target[target] = round(coverage, self.FINE_PRECISION)
```

Comment:
I would not round here. `self.FINE_PRECISION` is commented for reads per start point. Low pass sequencing will be rounded down to 0, which was an issue we had with the current Run Report. I am not sure what the rules for significant figures are for dividing two integers.

Reply:
Good point. Rule of thumb is to use the least number of significant figures in the inputs, which would imply 1 or 2 (given that depth of coverage is typically < 100). But easier to leave it unrounded I think.